### PR TITLE
Update rust-toolchain

### PR DESCRIPTION
**Pull request by gendx**
### Pull Request Overview

This pull request updates the Rust toolchain to today's nightly (nightly-2020-04-21).
- Fixed the `sed` commands in `tools/update_rust_version.sh` to work on Linux.
- Fixed some new warnings about unnecessary parentheses.
- Migrated from the now-deprecated `asm!` macro to its `llvm_asm!` replacement.

I've seen that the Tock 1.5 version is in progress (#1685), but almost no test have been done yet. It would be a nice opportunity to upgrade to a nightly as recent as possible for this release.

### Testing Strategy

This pull request was tested by Travis.

### TODO or Help Wanted

- [ ] It looks like this particular nightly triggers compiler crashes: https://github.com/rust-lang/rust/issues/70924.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

File: tools/update_rust_version.sh [hunk 1]
```diff
@@ -29,9 +29,9 @@ echo Updating Rust to $NIGHTLY
 echo $NIGHTLY > rust-toolchain
 
 # Update all relevant files with the new version string.
-sed -i ''  "s/nightly-[0-9]*-[0-9]*-[0-9]*/${NIGHTLY}/g" .travis.yml
-sed -i ''  "s/nightly-[0-9]*-[0-9]*-[0-9]*/${NIGHTLY}/g" .vscode/settings.json
-sed -i ''  "s/nightly-[0-9]*-[0-9]*-[0-9]*/${NIGHTLY}/g" doc/Getting_Started.md
-sed -i ''  "s/nightly-[0-9]*-[0-9]*-[0-9]*/${NIGHTLY}/g" rust-toolchain
-sed -i ''  "s/nightly-[0-9]*-[0-9]*-[0-9]*/${NIGHTLY}/g" tools/netlify-build.sh
-sed -i ''  "s/[0-9]*-[0-9]*-[0-9]*/${BEST_DATE}/g" shell.nix
+sed -i "s/nightly-[0-9]*-[0-9]*-[0-9]*/${NIGHTLY}/g" .travis.yml
```
Comment (ppannuto) [inline tools/update_rust_version.sh:32]: Ugh.

This is Mac / BSD `sed` versus Unix / coretools `sed`. On Mac, this change doesn't work:
```
$ sed -i "s/nightly-[0-9]*-[0-9]*-[0-9]*/${NIGHTLY}/g" .travis.yml
sed: 1: ".travis.yml": invalid command code .
```

I've had to fix this before, let me track down the solution and I'll push an update to this branch

**Comment by gendx** [comment]
Hum, this caused the following compiler crash on Travis.

```
thread 'rustc' panicked at 'failed to lookup `SourceFile` in new context', src/librustc_middle/ty/query/on_disk_cache.rs:456:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: rustc 1.44.0-nightly (20fc02f83 2020-04-20) running on x86_64-unknown-linux-gnu
note: compiler flags: -C opt-level=z -C debuginfo=2 -C debug-assertions=on -C incremental --crate-type lib
note: some of the compiler flags provided by cargo are hidden
```

I observed a similar issue at some point on my machine (https://github.com/rust-lang/rust/issues/71380).

So it looks like this particular nightly version is a bad one. Maybe we should keep the previous pinned version for the release.

**Comment by bradjc** [comment]
We normally update every two-ish months unless there is a problem, plus with this not passing travis, I say we continue with 1.5-rc1 to try to make this a timely release.

**Comment by ppannuto** [comment]
Upon second thinking, you're right, a last-minute nightly update doesn't really make sense if the goal is a stable release, but I do think this can merge right after the 1.5 freeze.

Re the build issue: That's a known cacheing problem, https://github.com/rust-lang/rust/issues/70924 -- clearing the travis cache fixed the build here, running clean locally should fix as well. Not sure whether that will be fixed or something people will just need to do as a one-off.

**Comment by gendx** [comment]
@ppannuto did someone manually re-run the Travis build? I wanted to reference it but now it unfortunately shows as "passed"...

**Comment by ppannuto** [comment]
I did, sorry

**Comment by ppannuto** [comment]
I updated the script to fix the `.bak` file problem

**Comment by bradjc** [comment]
Can we redo the build?

**Comment by gendx** [comment]
Marking as draft until nightly is usable again: https://github.com/rust-lang/rust/issues/70924

**Comment by hudson-ayers** [comment]
Separately from the issue that requires `cargo clean`, `make format` fails on this branch, throwing the error "failed to find targets"

**Comment by bradjc** [comment]
The blocking issue is only relevant to travis, correct? I don't think we can wait indefinitely for Rust to fix this. But hopefully this is patched soon.

I also just tried `make format` on today's nightly and it ran without issue.

**Comment by hudson-ayers** [comment]
Yeah, the `make format` issue was actually from an out of tree file getting picked up by the formatting script, sorry forgot to update here.

**Comment by gendx** [comment]
> The blocking issue is only relevant to travis, correct? I don't think we can wait indefinitely for Rust to fix this. But hopefully this is patched soon.

As discussed in #1654, this is unfortunately the trade-off of choosing Rust nightly over Rust stable. When some things break on nightly there is no prioritization to fix them within a day. And the bigger the codebase + the more unstable features we rely on, the likelier it is that at least one thing breaks in a nightly release (as we've already seen with nightly versions not supporting Rustfmt or Clippy).

In this particular case, we can get away by asking developers to explicitly `cargo clean` before each call to `cargo` (at the expense of longer compilation times) or whenever they see the compiler error.

On the other hand, after a month and a half lingering around, this bug is now "high priority", so i guess there is a chance it gets fixed in the next few weeks: https://github.com/rust-lang/rust/issues/70924#issuecomment-628092269.

**Comment by hudson-ayers** [comment]
https://github.com/rust-lang/rust/issues/70924 has been fixed! So if we update this to latest nightly we are probably good to go.

**Comment by gendx** [comment]
> [rust-lang/rust#70924](https://github.com/rust-lang/rust/issues/70924) has been fixed! So if we update this to latest nightly we are probably good to go.

It's probably easier to start from scratch to avoid merge conflicts from the past couple months. And the branch name is not valid anymore ;)

**Pull request closed without merging**